server/core/terraform/terraform_client.go

```diff
@@ -278,99 +277,86 @@ func (c *DefaultClient) TerraformBinDir() string {
-		matched := re.FindStringSubmatch(requiredVersionSetting)
+	if !c.downloadAllowed {
+		log.Debug("terraform downloads disabled.")
+		matched := c.ExtractExactRegex(log, requiredVersionSetting)
 		if len(matched) == 0 {
-			log.Debug("Did not specify exact version in terraform configuration, found %q", requiredVersionSetting)
+			log.Debug("did not specify exact version in terraform configuration, found %q", requiredVersionSetting)
+			return nil
+		}
+
+		version, err := version.NewVersion(matched[0])
+		if err != nil {
+			log.Err("error parsing version string: %s", err)
 			return nil
 		}
-		tfVersions = []string{matched[1]}
+		return version
 	}
 
-	constraint, _ := version.NewConstraint(requiredVersionSetting)
 	// Since terraform version 1.8.2, terraform is not a single file download anymore and
 	// Atlantis fails to download version 1.8.2 and higher. So, as a short-term fix,
 	// we need to block any version higher than 1.8.1 until proper solution is implemented.
 	// More details on the issue here - https://github.com/runatlantis/atlantis/issues/4471
-	highestSupportedConstraint, _ := version.NewConstraint("<= 1.8.1")
-	versions := make([]*version.Version, len(tfVersions))
-
-	for i, tfvals := range tfVersions {
-		newVersion, err := version.NewVersion(tfvals)
-		if err == nil {
-			versions[i] = newVersion
-		}
+	constraintStr := fmt.Sprintf("%s, <= 1.8.1", requiredVersionSetting)
+	vc, err := version.NewConstraint(constraintStr)
+	if err != nil {
+		log.Err("Error parsing constraint string: %s", err)
+		return nil
 	}
 
-	if len(versions) == 0 {
-		log.Debug("Did not specify exact valid version in terraform configuration, found %q", requiredVersionSetting)
+	constrainedVersions := &releases.Versions{
+		Product:     product.Terraform,
+		Constraints: vc,
+	}
+	installCandidates, err := constrainedVersions.List(context.Background())
+	if err != nil {
+		log.Err("error listing available versions: %s", err)
+		return nil
+	}
+	if len(installCandidates) == 0 {
+		log.Err("no Terraform versions found for constraints %s", constraintStr)
 		return nil
 	}
 
-	sort.Sort(sort.Reverse(version.Collection(versions)))
+	// We want to select the highest version that satisfies the constraint.
+	versionDownloader := installCandidates[len(installCandidates)-1]
 
-	for _, element := range versions {
-		if constraint.Check(element) && highestSupportedConstraint.Check(element) { // Validate a version against a constraint
-			tfversionStr := element.String()
-			if lib.ValidVersionFormat(tfversionStr) { //check if version format is correct
-				tfversion, _ := version.NewVersion(tfversionStr)
-				log.Info("Detected module requires version: %s", tfversionStr)
-				return tfversion
-			}
-		}
-	}
-	log.Debug("Could not match any valid terraform version with %q", requiredVersionSetting)
-	return nil
+	// Get the Version object from the versionDownloader.
+	downloadVersion := versionDownloader.(*releases.ExactVersion).Version
```

Comment:
I think the main reason would be future compatibility (adding in support for OpenTofu for example).

DetectVersion is called [here](https://github.com/runatlantis/atlantis/blob/3f87f7b5d7c7d3d4460f5fab1b9230a7326a1836/server/events/project_command_context_builder.go#L124-L128) and [here](https://github.com/runatlantis/atlantis/blob/3f87f7b5d7c7d3d4460f5fab1b9230a7326a1836/server/events/project_command_context_builder.go#L177-L181) to determine the version if not explicitly set in the Atlantis config. The field that is populated is a `*version.Version` field, which would be compatible with either Terraform or OpenTofu.

If we switched it to be a src.Source field, then it wouldn't be compatible with OpenTofu moving forward. For any future work on supporting OpenTofu, an additional field (say, `TofuVersion`) may be added so it may not matter what type this `TerraformVersion` field is. But for the sake of possibility, I think keeping it as-is is the best cause of action.

I don't think it's unreasonable to want to know the version that will be installed beforehand without running it
- Logging purposes - may wish to log what version was selected
- Security reasons - shouldn't have to run a binary to identify the version - this is heightened by the binary downloaded just being called `terraform` (with no version number) from what I can gather running hc-install locally. Is there a way to customise the binary name? Please correct me if I am wrong!
- Choices - whilst not the case in this repo, there may be scenarios where a user/application may want to make choices based on the version number. e.g. if greater than x, alert user that feature y is not included.

Reply:
Gotcha, yeah that makes sense - I appreciate you looking into it and giving a detailed reply!